### PR TITLE
Factored out product-create, product-purchase, and user-info apis

### PR DESCRIPTION
**Pull request by laurwang**
Made a single api for writing events to kinesis and removed the three apis we had that were writing to kinesis separately.  (Have put this api in the retail-stream service for the moment, but the consensus seems to be too split this out to a separate service.)  This PR also includes changes to the web-app to use the single api, and has the branch of PR #54 merged in, so is set up assuming all services are in the same account.  Conflicts with PR #54 have been fixed; these were to do with the deletion of the three apis.  In addition, have removed code around the CatalogReaderRole from the web-app to complete the transfer of the ProductCatalogRole's rights to the WebApplicationIdentity role.

To summarize, apart from the changes due to PR #54 , this PR does not affect the product-catalog and product-photo services.  The main changes are to the retail-stream and web services, with the removal of the product-create, product-purchase, and user-info services.

TODO: move out the work by the single api for validation to a validation and schema registration service.  The schemas are currently in an array set in the lambda (retailStreamApi.js) that backs up the single api.

### CODE REVIEW

File: web/components/login/amazon-login.jsx [hunk 1]
```diff
@@ -169,10 +167,11 @@ class AmazonLogin extends Component {
   }
 
   sendUserLogin() {
-    this.makeApiRequest(config.UserInfoAPI, 'POST', '/login-user/', {
+    this.makeApiRequest(config.RetailStreamAPI, 'POST', '/event-writer/', {
       schema: 'com.nordstrom/user-info/login/1-0-0',
       id: this.state.profile.id,
       name: this.state.profile.name,
+      origin: `hello-retail/web-client-login-user/${this.state.profile.email}/${this.state.profile.name}`,
```
Comment (gwsii) [inline web/components/login/amazon-login.jsx]: This is an example of kind of data we might want to discuss before putting on the envelope. A person's e-mail and name are PII.

**Comment by erikerikson** [comment]
perhaps this was intended to target the `develop-remove-cross-account-patterns` branch rather than `develop`?

File: package.json [hunk 2]
```diff
@@ -77,9 +69,11 @@
     "web:build:test": "pushd ./web/ && sls buildWebApp -s test",
     "web:build:prod": "pushd ./web/ && sls buildWebApp -s prod",
     "web:build:demo": "pushd ./web/ && sls buildWebApp -s demo",
+    "web:build:bimx": "pushd ./web/ && sls buildWebApp -s bimx",
     "web:deploy:test": "pushd ./web/ && sls deploy -s test && popd",
     "web:deploy:prod": "pushd ./web/ && sls deploy -s prod && popd",
     "web:deploy:demo": "pushd ./web/ && sls deploy -s demo && popd",
+    "web:deploy:bimx": "pushd ./web/ && sls deploy -s bimx && popd",
     "web:dev:hot": "./node_modules/.bin/webpack-dev-server --config ./web/webpack.config.js --https --hot --progress --colors --watch --display-error-details --display-cached --content-base ./web/app/"
```
Comment (laurwang) [inline package.json]: Have pushed that in just now

File: web/components/login/amazon-login.jsx (hunk 1 above)
Comment (laurwang) [inline web/components/login/amazon-login.jsx]: Yes, definitely need a plan for PII, not addressed in this PR though.

File: retail-stream/retailStreamApi.js [hunk 3]
```diff
@@ -0,0 +1,180 @@
+'use strict'
```
Comment (erikerikson) [inline retail-stream/retailStreamApi.js]: It seems we had different expectations about how this API would be factored.  Let's discuss this.

File: retail-stream/retailStreamApi.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+const productCreateSchemaId = makeSchemaId(productCreateSchema)
+const userLoginSchemaId = makeSchemaId(userLoginSchema)
+const updatePhoneSchemaId = makeSchemaId(updatePhoneSchema)
+const addRoleSchemaId = makeSchemaId(addRoleSchema)
+ajv.addSchema(productPurchaseSchema, productPurchaseSchemaId)
+ajv.addSchema(productCreateSchema, productCreateSchemaId)
+ajv.addSchema(userLoginSchema, userLoginSchemaId)
+ajv.addSchema(updatePhoneSchema, updatePhoneSchemaId)
+ajv.addSchema(addRoleSchema, addRoleSchemaId)
+const schema = [
```
Comment (erikerikson) [inline retail-stream/retailStreamApi.js]: schemas?

File: retail-stream/retailStreamApi.js [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+  success: items => impl.response(200, JSON.stringify(items)),
+
+  checkKnownSchema: (eventData) => {
+    const results = {
+      errors: [],
+      matches: [], // TODO does ajv actually check the content of the schema field vs the eventData schema field or just that it is of the format url?  if latter, then could have multiple matches.
+    }
+
+    schema.forEach((scheme) => {
```
Comment (erikerikson) [inline retail-stream/retailStreamApi.js]: `scheme` => `schema` after `schema` => `schemas`?

File: retail-stream/serverless.yml [hunk 6]
```diff
@@ -9,12 +9,13 @@ custom:
...
   retailStreamVersion: 1 # The version of this service to depend upon
-  retailStream: ${self:} # externally, use `${file(~/retail-stream/serverless.yml)}`, replacing the `~` therein with the appropriate relative reference [e.g. '..' or '../..']  userInfoApiVersion: 1
+  retailStream: ${self:} # externally, use `${file(~/retail-stream/serverless.yml)}`, replacing the `~` therein with the appropriate relative reference [e.g. '..' or '../..']
   # Exports
   exportRetailStreamName: ${self:custom.retailStream.service}:${self:custom.stage}:RetailStreamName:${self:custom.retailStreamVersion}
   exportRetailStreamArn: ${self:custom.retailStream.service}:${self:custom.stage}:RetailStreamArn:${self:custom.retailStreamVersion}
   exportRetailStreamWriterArn: ${self:custom.retailStream.service}:${self:custom.stage}:RetailStreamWriterArn:${self:custom.retailStreamVersion}
   exportRetailStreamReaderArn: ${self:custom.retailStream.service}:${self:custom.stage}:RetailStreamReaderArn:${self:custom.retailStreamVersion}
+  exportRetailStreamRestAPI: ${self:custom.retailStream.service}:${self:custom.stage}:RetailStreamRestAPI:${self:custom.retailStreamVersion}
```
Comment (erikerikson) [inline retail-stream/serverless.yml]: `..stAPI` in camel case is `...stApi`

File: retail-stream/serverless.yml [hunk 7]
```diff
@@ -25,15 +26,46 @@ provider:
   environment:
     STAGE: ${self:custom.stage}
 
+functions:
```
Comment (erikerikson) [inline retail-stream/serverless.yml]: Another difference in expectation was that specific endpoints would be provided per website interaction type in order to better collect the declared interaction of the website with the backend.  i.e. that there wouldn't be a single submission point for events

File: retail-stream/serverless.yml [hunk 8]
```diff
@@ -114,3 +190,9 @@ resources:
         Fn::GetAtt: [ RetailStreamReader, Arn ]
       Export:
         Name: ${self:custom.retailStream.custom.exportRetailStreamReaderArn}
+    RetailStreamRestAPI:
```
Comment (erikerikson) [inline retail-stream/serverless.yml]: `...stAPI` => `...stApi`

File: retail-stream/serverless.yml [hunk 9]
```diff
@@ -114,3 +190,9 @@ resources:
...
+      Description: API used to submit producer events to the Retail Stream
+      Value:
+        Ref: ApiGatewayRestApi
+      Export:
+        Name: ${self:custom.retailStream.custom.exportRetailStreamRestAPI}
```
Comment (erikerikson) [inline retail-stream/serverless.yml]: `...stAPI` => `...stApi`

File: web/serverless.yml [hunk 10]
```diff
@@ -24,7 +24,7 @@ custom:
       - DomainName
 
   # Hosted Zone ID for all of CloudFormation.
-  cloudFrontHostedZoneId: Z2FDTNDATAQYW2
+  cloudFrontHostedZoneId: ${self:custom.private.hostedZoneId}
```
Comment (erikerikson) [inline web/serverless.yml]: please delete, replacing all references to `${self:custom.private.cloudFrontHostedZoneId}` with `${self:custom.private.hostedZoneId}`

File: retail-stream/serverless.yml [hunk 11]
```diff
@@ -9,12 +9,13 @@ custom:
   # Retail Stream Versions & Exports
   # WARNING: each project that uses these will have to define the following two attributes in its serverless.yml: (custom.retailStreamVersion, custom.retailStream)
   retailStreamVersion: 1 # The version of this service to depend upon
-  retailStream: ${self:} # externally, use `${file(~/retail-stream/serverless.yml)}`, replacing the `~` therein with the appropriate relative reference [e.g. '..' or '../..']  userInfoApiVersion: 1
+  retailStream: ${self:} # externally, use `${file(~/retail-stream/serverless.yml)}`, replacing the `~` therein with the appropriate relative reference [e.g. '..' or '../..']
```
Comment (erikerikson) [inline retail-stream/serverless.yml]: thanks!

File: retail-stream/retailStreamApi.js (hunk 5 above)
Comment (erikerikson) [inline retail-stream/retailStreamApi.js]: also, the `eventData` (as per line 134 below) declares the schema to which it adheres.  Why not check that we have a schema defined that has a matching ID with the schema ID that is declared and then check against the declared schema rather than cycling through all potential schemas?

File: web/serverless.yml (hunk 10 above)
Comment (laurwang) [inline web/serverless.yml]: ok

File: retail-stream/retailStreamApi.js (hunk 4 above)
Comment (laurwang) [inline retail-stream/retailStreamApi.js]: ok

**Comment by laurwang** [comment]
Have changed validation of schema to demand that the passed data from the caller have a valid schema name and am checking just against the one declared (rather than allowing anything that matched a registered schema).

Have removed the eventWriter resources from retail-stream and deployed as separate service, event-writer-api.

File: event-writer/eventWriterApi.js [hunk 12]
```diff
@@ -102,32 +59,35 @@ const impl = {
...
+          }),
+          PartitionKey: eventData.id, // TODO if some schema use id field something other than the partition key, the schema need to have a keyName field and here code should be eventData[eventData.keyName]
+          StreamName: process.env.STREAM_NAME,
         }
-      })
+
+        kinesis.putRecord(newEvent, (err, data) => {
+          if (err) {
+            callback(null, impl.kinesisError(eventData.schema, err))
+          } else if (data) {
```
Comment (erikerikson) [inline event-writer/eventWriterApi.js]: if err is falsy, then data is guaranteed to be valid, as per the api contract with kinesis.  with `if (data)` check is unnecessary

File: event-writer/eventWriterApi.js [hunk 13]
```diff
@@ -63,37 +41,16 @@ const impl = {
     body,
   }),
 
-  clientError: (errors, event) => impl.response(400,
-    `${constants.API_NAME} ${constants.INVALID_REQUEST} could not validate request to any known event schema. Errors by schema: '${JSON.stringify(errors)}'.  Found in event: '${JSON.stringify(event)}'`),
+  clientError: (error, event) => impl.response(400,
+    `${constants.API_NAME} ${constants.INVALID_REQUEST}  ${error}.  Event: '${JSON.stringify(event)}'`),
 
-  kinesisError: (methodNames, err) => {
+  kinesisError: (schemaName, err) => {
```
Comment (erikerikson) [inline event-writer/eventWriterApi.js]: Not terribly important but there's a discussion we haven't thoroughly had as a team about the verbosity of external messages.  Sometimes, it's best to remark "ServerError" and not leak system information (an example of this is in `~/product-photos/3.receive/receive.js`).

File: event-writer/eventWriterApi.js [hunk 14]
```diff
@@ -102,32 +59,35 @@ const impl = {
...
+          PartitionKey: eventData.id, // TODO if some schema use id field something other than the partition key, the schema need to have a keyName field and here code should be eventData[eventData.keyName]
+          StreamName: process.env.STREAM_NAME,
         }
-      })
+
+        kinesis.putRecord(newEvent, (err, data) => {
+          if (err) {
+            callback(null, impl.kinesisError(eventData.schema, err))
+          } else if (data) {
+            callback(null, impl.success(`Event with schema ${eventData.schema}: ${JSON.stringify(data)}`))
```
Comment (erikerikson) [inline event-writer/eventWriterApi.js]: do we want to place all the data going onto the stream also into the log?  It may be sufficient to simply `callback()` as a means of completing successfully

File: web/serverless.yml [hunk 15]
```diff
@@ -54,9 +54,12 @@ custom:
   private: ${file(../private.yml)}
 
   helloRetail: ${file(../helloRetail.yml)}
-
+  # TODO Suspect some strangeness in referencing of external serverless files by productCatalogBuilder requires these be provided
```
Comment (erikerikson) [inline web/serverless.yml]: Yes, currently a `serverless.yml` that imports/depends on another `serverless.yml` must import all dependencies of the imported `serverless.yml`.  This is unpleasant but necessary for the shorter term.  This should be resolved when this issue is resolved: https://github.com/serverless/serverless/issues/3233

Comment (erikerikson) [inline web/serverless.yml]: It may also be impacting the `eventWriterApi` service since it will have the `retailStream` service as a dependency in order to provide itself rights to `putRecord` onto the stream (lines 8, 9, and 94 of the serverless.yml).

File: event-writer/eventWriterApi.js (hunk 13 above)
Comment (laurwang) [inline event-writer/eventWriterApi.js]: Yes, agree we need a discussion.  Would be great to have a module we use consistently for errors.  Going to leave this as is, till we have a plan.

File: event-writer/eventWriterApi.js (hunk 12 above)
Comment (laurwang) [inline event-writer/eventWriterApi.js]: Agree, taking it out

File: web/serverless.yml (hunk 15 above)
Comment (laurwang) [inline web/serverless.yml]: Thanks for the ticket on this issue.

File: event-writer/eventWriterApi.js (hunk 14 above)
Comment (laurwang) [inline event-writer/eventWriterApi.js]: will just return the response from kinesis, instead of the verbose message, but need to use impl.success for the status code.

File: event-writer/serverless.yml [hunk 16]
```diff
@@ -49,63 +49,54 @@ functions:
...
-        RoleName: ${self:custom.stage}ProductCreationWriter${self:custom.productCreationApiVersion}
+        RoleName: ${self:custom.stage}EventWriter${self:custom.eventWriterApiVersion}
         AssumeRolePolicyDocument:
           Version: '2012-10-17'
           Statement:
             - Effect: Allow
               Action: sts:AssumeRole
               Principal:
-                Service: lambda.amazonaws.com
+                Service: 'lambda.amazonaws.com'
```
Comment (gwsii) [inline event-writer/serverless.yml]: Interesting that you re-added the quotes. Did it turn out that they are needed?

File: event-writer/serverless.yml [hunk 17]
```diff
@@ -49,63 +49,54 @@ functions:
...
         ManagedPolicyArns:
           - ${self:custom.private.teamPolicy}
         Policies:
           - PolicyName: CreateAndWriteToLogStream
             PolicyDocument:
               Version: '2012-10-17'
               Statement:
-                - Effect: Allow
+                - Effect: 'Allow'
```
Comment (erikerikson) [inline event-writer/serverless.yml]: also unnecessary quotes here

File: event-writer/serverless.yml [hunk 18]
```diff
@@ -49,63 +49,54 @@ functions:
...
                       - ':'
-                      - - Fn::GetAtt: [ ProductCreationLogGroup, Arn ]
+                      - - Fn::GetAtt: [ EventWriterLogGroup, Arn ]
                         - '*'
           - PolicyName: WriteToKinesis
             PolicyDocument:
               Version: '2012-10-17'
               Statement:
-                - Effect: Allow
+                - Effect: 'Allow'
```
Comment (erikerikson) [inline event-writer/serverless.yml]: and here

File: web/serverless.yml (hunk 15 above)
Comment (erikerikson) [inline web/serverless.yml]: mind removing this comment?
